File: solutions/0239-sliding-window-maximum/solution.cpp
class Solution {
public:
    vector<int> maxSlidingWindow(vector<int>& nums, int k) {
        deque<pair<int,int>> dq;
        vector<int> ans;
        for(int right=0;right<nums.size();right++){
            //remove invalids
            while(!dq.empty() and (dq.front().first)<=(right-k)){
                dq.pop_front();
            }
            //remove until smaller elements from end
            int curr = nums[right];
            while(!dq.empty() and dq.back().second<=curr){
                dq.pop_back();
            }
            //insert the curr
            dq.push_back({right,curr});

            //push in asnwer
            if(right>=k-1){
                ans.push_back(dq.front().second);
            }
        }
        return ans;
    }
};

File: solutions/0239-sliding-window-maximum/solution.py
class Solution:
    def maxSlidingWindow(self, nums: List[int], k: int) -> List[int]:
        from collections import deque
        dq = deque()
        ans = []
        def insertInDq(item,i):
            if not dq:
                dq.append((item,i))
                return
            while dq and dq[-1][0]<=item:
                dq.pop()
            dq.append((item,i))

        for i in range(0,len(nums)):
            insertInDq(nums[i],i)
            # for taking max, make sure it is valid in range
            if i>=k-1:
                while dq and dq[0][1]<i-k+1:
                    dq.popleft()
                ans.append(dq[0][0])
        return ans





        
